include/tvm/ir/instrument.h

```diff
@@ -0,0 +1,114 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+/*!
+ * \file tvm/ir/instrument.h
+ *
+ * This file introduces a pass instrument infrastructure, inspired from LLVM and MLIR.
+ * It inserts instrumentation points between passes run.
+ *
+ * Within a pass context (tvm::transfom::PassContext), the instrumentation call sequence will like:
+ *
+ *   Instrument SetUp
+ *
+ *     if (Instrument Before Pass)
+ *       Pass Run
+ *       Instrument After Pass
+ *
+ *     if (Instrument Before Pass)
+ *       Pass Run
+ *       Instrument After Pass
+ *
+ *   Instrument TearDown
+ *
+ *
+ * Instrument point before pass can determine particular pass is disable or not depends on the
+ * callback registered.
+ */
+#ifndef TVM_IR_INSTRUMENT_H_
+#define TVM_IR_INSTRUMENT_H_
+
+#include <tvm/node/reflection.h>
+#include <tvm/runtime/container.h>
+
+#include <utility>
+#include <vector>
+
+namespace tvm {
+
+class IRModule;
+
+// Forward class for PassInstrumentNode methods
+namespace transform {
+class PassInfo;
+}  // namespace transform
+
+namespace instrument {
+
+/*!
+ * \brief PassInstrumentNode forms an instrument implementation.
+ * It provides API for users to register callbacks at different instrument point.
+ * \sa PassInstrument
+ */
+class PassInstrumentNode : public Object {
+ public:
+  virtual ~PassInstrumentNode() {}
+
+  /*! \brief Set up environment for instrumentation. */
+  virtual void SetUp() const = 0;
+
+  /*! \brief Clean up instrumentation environment. */
+  virtual void TearDown() const = 0;
+
+  /*!
+   * \brief Instrument before pass run, determine whether to run the pass or not.
+   * \param mod The module that an optimization pass runs on.
+   * \param info The pass information.
+   *
+   * \return true to run the pass; false to skip the pass.
+   */
+  virtual bool RunBeforePass(const IRModule& mod, const transform::PassInfo& info) const = 0;
```

Comment:
Can calls to `RunBeforePass` be nested?

Reply:
No. I think.
Like the ```InstrumentBeforePass``` logic before.

Reply:
Well, passes can be nested, so `RunBeforePass` needs to be able to handle nested calls.